### ISSUE TITLE
vctrs

### ISSUE THREAD

**Issue by thebioengineer** [thebioengineer/colortable#4]
refactor to vctrs? seems like it might be easier for sustainability, offload some maintenance tools.

**Comment by thebioengineer** [comment]
Working on implementing in the investigate_vctrs branch (#11 )

@krlmlr - I have a question regarding how to reach my goal with colortable using vctrs. 

My goal is that the color_vctr object is simply a wrapper for providing formatting options that are able to be applied across printing methods (console, pdf, html, etc). Otherwise the behavior of the input vector is the same as before. 

Right now my plan is to go about implementing the various methods - vec_math, vec_proxy, etc. such that the internal vector gets called, the manipulation, etc, happens. and it gets re-wrapped into color_vctr. Would you suggest a different approach?

**Comment by krlmlr** [comment]
Maybe you can get away with implementing `vec_restore()` for your class?

**Comment by thebioengineer** [comment]
I am currently using a vctr_rcrd since styling happens independently for each record, so it isnt a case of using attributes. Unless I misunderstand?

**Comment by krlmlr** [comment]
I wasn't aware of that. Composition (=using a record) means that you have to forward all operations that you want to make available. Inheritance (=using attributes to store the extra data) means you "only" need to restore and be careful with slicing. The former is often the safer approach -- needs a bit of boilerplate code though.

Since you already have an implementation, I'd suggest to keep it unless it turns out to be impractical.

Do you offer an easy way, at the user level, to isolate the underlying data and the styling attributes, and to recombine?

**Comment by thebioengineer** [comment]
At this moment I don't have tools outside of what is in {vctrs} to handle the fields/etc. My original implementation without vctrs could do that. I had methods for slicing (`[`) but vctrs appears to ignore any methods like that in favor of vctr_slice once it goes into tidyverse functions. Which is why I moved to the vctrs version so I could have it play nicely with the tidyverse tools.

**Comment by thebioengineer** [comment]
It is looking like this is NECESSARY to work with the tidyverse now. Upping in priority!

**Comment by krlmlr** [comment]
Why is it now necessary?

**Comment by thebioengineer** [comment]
oh, sorry @krlmlr , didn't think this would tag you! but the below code is why

```r
library(colortable)
library(tidyverse)

set.seed(2020)
x <- tibble(x = runif(10))

## works
x$vec_1 <- color_vctr(1:10, text_color = "blue")

x

## does not error out, but attributes are not assigned
x <- x %>% 
  mutate(
    vec_2 = color_vctr(1:10, text_color = "blue")
  )

#but printing will not be done correctly b/c missing attribs
x

```

even pre-defining the color_vctr does not work

```r
library(colortable)
library(tidyverse)

set.seed(2020)
x <- tibble(x = runif(10))

c_vect <- color_vctr(1:10, text_color = "orange")

c_vect
## does not error out, but attributes are not assigned
x <- x %>% 
  mutate(
    vec_1 = c_vect
  )
#but printing it will cause an error b/c no attribs
x
```

**Comment by krlmlr** [comment]
I'd say you'll need less custom code if you use vctrs, I'm pretty sure we *could* also make it work without vctrs (because nothing in dplyr checks if the data is a vctrs data type).

**Comment by thebioengineer** [comment]
Moving over to vctrs has been the long term plan once I figured out how to use it! I have a branch with my initial experimentation. I needed a vctr_rcrd to get everything to behave but at the time dplyr(mutate) did not support vctr_rcrds. With the release of 1.0.0 my understanding was that it would.

**Comment by thebioengineer** [comment]
currently resolved in #23 - questions remain about how to make it clear that we only want to be a wrapper where the class controls printing and not "internal" class behavior. Lionel and Hadley conversing on r-lib/vctrs#1080

**Comment by thebioengineer** [comment]
performed merge to master.
